packages/datadog-plugin-undici/src/index.js

```diff
@@ -0,0 +1,291 @@
+    const uri = `${request.origin}${path}`
+
+    const span = asyncLocalStorage.getStore()
+    if (span) {
+      span.addTags({
+        'resource.name': method,
+        'span.type': 'http',
+        'http.method': method,
+        'http.url': uri,
+        'service.name': getServiceName(tracer, config, request.origin)
+      })
+      requestSpansMap.set(request, span)
+    }
+
+    const headers =
+      typeof request.headers === 'string'
+        ? parseHeaders(request.headers)
+        : request.headers
+
+    if (
+      !(
+        hasAmazonSignature({ ...request, headers }) ||
+        !config.propagationFilter(uri)
+      )
+    ) {
+      const injectedHeaders = {}
+      tracer.inject(span, HTTP_HEADERS, injectedHeaders)
+      Object.entries(injectedHeaders).forEach(([key, value]) => {
+        request.addHeader(key, value)
+      })
+    }
+
+    analyticsSampler.sample(span, config.measured)
+  }
+
+  function handleRequestError ({ request, error }) {
+    const span = requestSpansMap.get(request)
+    addErrorToSpan(span, error)
+    addRequestHeaders(request, span, config)
+    span.finish()
+  }
+
+  function handleRequestHeaders ({ request, response }) {
+    const span = requestSpansMap.get(request)
+    addRequestHeaders(request, span, config)
+    setStatusCode(response, span, config)
+    config.hooks.request(span, request, response)
+  }
+
+  return function unsubscribe () {
+    if (channels.requestChannel.hasSubscribers) {
+      channels.requestChannel.unsubscribe(handleRequestCreate)
+    }
+    if (channels.headersChannel.hasSubscribers) {
+      channels.headersChannel.unsubscribe(handleRequestHeaders)
+    }
+    if (channels.errorChannel.hasSubscribers) {
+      channels.errorChannel.unsubscribe(handleRequestError)
+    }
+  }
+}
+
+function addRequestHeaders (req, span, config) {
+  const headers = parseHeaders(req.headers)
+  Object.entries(headers).forEach(([key, value]) => {
+    span.setTag(`${HTTP_REQUEST_HEADERS}.${key}`, value)
+  })
+
+  if (!headers.host) {
+    // req.servername holds the value of the host header
+    if (req.servername) {
+      span.setTag(`${HTTP_REQUEST_HEADERS}.host`, req.servername)
+    } else {
+      // Undici's host header are written directly
+      // to the stream, and not passed to the `Request` object
+      // This workaround ensure we set the host if
+      // it was not explicitely provided
+      const { hostname, port } = url.parse(req.origin)
+      const host = `${hostname}${port ? `:${port}` : ''}`
+      span.setTag(`${HTTP_REQUEST_HEADERS}.host`, host)
+    }
+  }
+}
+
+function setStatusCode (res, span, config) {
+  // fetch has status set on `status` rather than statusCode
+  const statusCode = res.status || res.statusCode
+  span.setTag(HTTP_STATUS_CODE, statusCode)
+
+  if (!config.validateStatus(statusCode)) {
+    span.setTag('error', 1)
+  }
+}
+
+function addResponseHeaders (res, span, config) {
+  config.headers.forEach((key) => {
```

Comment:
A for...of is significantly faster.

Reply:
Thanks for the tip! Does a shared resource exist for such optimisations as an Eslint rule set for example?